File: ProblemSet/427. 生成括号/java/427. 生成括号 长度递归穷举法.java
import java.util.List;
import java.util.ArrayList;

public class Solution {
    private List<String>[] memo;
    /**
     * @param n: n pairs
     * @return: All combinations of well-formed parentheses
     */
    public List<String> generateParenthesis(int n) {
        memo = new ArrayList[n + 1];
        return generate(n);
    }

    private List<String> generate(int n) {
        if (memo[n] != null) {
            return memo[n];
        }
        List<String> ans = new ArrayList<String>();
        if (n == 0) 
            ans.add("");
        else {
            for (int i = 0; i < n; ++i) {
                for (String a : generate(i)) {
                    for (String b : generate(n - i - 1))
                        ans.add("(" + a + ")" + b);
                }
            }
        }
        memo[n] = ans;
        return ans;
    }
}

File: ProblemSet/427. 生成括号/python/427. 生成括号 长度递归穷举法.py
from functools import lru_cache

class Solution:
    """
    @param n: n pairs
    @return: All combinations of well-formed parentheses
    """
    def generateParenthesis(self, n):
        @lru_cache(n + 1)
        def generate(n):
            if n == 0:
                return ['']
            ans = []
            for i in range(n):
                for a in generate(i):
                    for b in generate(n - i - 1):
                        ans.append('({}){}'.format(a, b))
            return ans
        
        return generate(n)
